Virtualbox: failed to load libGL.so.1: VERR_FILE_NOT_FOUND

### Nixpkgs version

- Stable (24.11)

### Describe the bug

I can't run `sway` in Virtualbox. I believe the problem is not `sway` per se, but rather related to 3d acceleration not working. I have been able to run the below `configuration.nix` successfully (=shows `glxgears` after boot) with `qemu` but not with Virtualbox. For Virtualbox, the graphical output is:

![Image](https://github.com/user-attachments/assets/3a049b0e-70e0-42e4-a184-375c456bc7fd)

```console
VMware: No 3D enabled (0, Success).
libEGL warning: egl: failed to create dri2 screen
00:00:00.040 [ERROR] [wlr] [EGL] command: eglInitialize, error: EGL_NOT_INITIALIZED (0x3001), message: "DRI2: failed to create screen"
VMware: No 3D enabled (0, Success).
libEGL warning: egl: failed to create dri2 screen
00:00:00.046 [ERROR] [wlr] [EGL] command: eglInitialize, error: EGL_NOT_INITIALIZED (0x3001), message: "DRI2: failed to create screen"
VMware: No 3D enabled (0, Success).
libEGL warning: egl: failed to create dri2 screen
00:00:00.053 [ERROR] [wlr] [EGL] command: eglInitialize, error: EGL_NOT_INITIALIZED (0x3001), message: "DRI2: failed to create screen"
00:00:00.053 [ERROR] [wlr] [EGL] command: eglInitialize, error: EGL_NOT_INITIALIZED (0x3001), message: "eglInitialize"
00:00:00.053 [ERROR] [wlr] [render/egl.c:268] Failed to initialize EGL
00:00:00.053 [ERROR] [wlr] [render/egl.c:571] Failed to initialize EGL context
00:00:00.053 [ERROR] [wlr] [render/gles2/renderer.c:499] Could not initialize EGL
00:00:00.053 [ERROR] [wlr] [render/wlr_renderer.c:272] Could not initialize renderer
00:00:00.053 [ERROR] [sway/server.c:236] Failed to create renderer
```

Looking into the logs from Virtualbox reveals:

```
00:00:02.213596 rtldrNativeLoad: dlopen('libGL.so.1', RTLD_NOW | RTLD_LOCAL) failed: libGL.so.1: cannot open shared object file: No such file or directory
00:00:02.227553 VMSVGA3d: Error loading OpenGL library and resolving necessary functions: VERR_VGA_GL_SYMBOL_NOT_FOUND
00:00:02.227557 VMSVGA3d: 3D support disabled! (vmsvga3dInit -> VERR_VGA_GL_SYMBOL_NOT_FOUND)
00:00:01.899286 VMSVGA3d: failed to load libGL.so.1: VERR_FILE_NOT_FOUND
00:00:01.899292 AssertLogRel /build/VirtualBox-7.1.4/src/VBox/Devices/Graphics/DevVGA-SVGA3d-glLdr.cpp(328) int glLdrInit(PPDMDEVINS): pfnRet
00:00:01.899296 glXGetFBConfigAttrib missing
00:00:01.899300 VMSetError: /build/VirtualBox-7.1.4/src/VBox/Devices/Graphics/DevVGA-SVGA3d-glLdr.cpp(328) int glLdrInit(PPDMDEVINS); rc=VERR_VGA_GL_SYMBOL_NOT_FOUND
00:00:01.899302 VMSetError: Missing OpenGL symbol 'glXGetFBConfigAttrib'
00:00:01.899303 
00:00:01.899310 VMSVGA3d: Error loading OpenGL library and resolving necessary functions: VERR_VGA_GL_SYMBOL_NOT_FOUND
00:00:01.899314 VMSVGA3d: 3D support disabled! (vmsvga3dInit -> VERR_VGA_GL_SYMBOL_NOT_FOUND)
```

I believe this to be the same problem with any combination of `virtualisation.virtualbox.host.package = pkgs.virtualbox[Hardened]` and/or `virtualisation.virtualbox.host.enableHardening = [true|false];`

I confirmed that my setup is able to run KDE/Plasma (see `configuration.nix` below).

I confirmed that my setup is able to run sway witin Qemu.

The libGL error is obviously the same but Plasma is able to launch a graphical user environment. This could of course mean that the problem is *not* libGL/Virtualbox but sway. I also tested `cage` with no luck neither (similar error message), same for Hyprland (which aborts with this [error](https://github.com/hyprwm/Hyprland/blob/v0.45.2/src/render/OpenGL.cpp#L114)). 

### Steps to reproduce

I used this `configuration.nix` and `nix shell nixpkgs/nixos-24.11#nixos-generators --command nixos-generate -f virtualbox -c ./configuration.nix`:

```nix
{ config, pkgs, self, ... }: {
  # Either ...
  programs.sway = { enable = true; };
  environment.variables = {
    WLR_NO_HARDWARE_CURSORS = 1;
  }; # https://www.reddit.com/r/swaywm/comments/127fsxr/
  services.getty.autologinUser = "root";
  programs.bash.loginShellInit = ''
    if [ "$(tty)" = "/dev/tty1" ]; then
      sway --config ${
        pkgs.writeText "glxgears" ''
          exec ${pkgs.mesa-demos}/bin/glxgears -fullscreen
        ''
      }
    fi
  '';

  # ... or
  # services.displayManager.sddm.enable = true;
  # services.displayManager.sddm.wayland.enable = true;
  # services.desktopManager.plasma6.enable = true;

  virtualisation.virtualbox.guest.enable = true;
}
```
(~1.6Gb .ova file, 2.8Gb for KDE ...)

Then, import the appliance using File -> Import Appliance. Before launching, open the Settings -> Display -> Enable 3d Acceleartion. Launch the VM, open the Log (Machine -> Show Log...)


### Expected behaviour

No error message about a missing libGL.so.1 in the logs from Virtualbox. `sway` starts up and can show `glxgears`

### Screenshots

see above

### Relevant log output

see above

### Additional context

_No response_
Some resources that are relevant IMHO:

* https://github.com/NixOS/nixpkgs/issues/22760
* https://github.com/NixOS/nixpkgs/pull/279009https://www.reddit.com/r/swaywm/comments/127fsxr/sway_would_not_start_installed_in_virtualbox/
* https://github.com/NixOS/nixpkgs/pull/313390
* https://github.com/NixOS/nixpkgs/pull/60985 (the introduced shell function/package has been renamed from `addOpenGLRunpath` to addDriverRunPath` in the meantime)
* https://github.com/NixOS/nixpkgs/pull/378464 (silly me!)
* https://github.com/NixOS/nixpkgs/pull/279009 (unrelated to virtualbox but may give some insights, though)

```console
[I] rolf@tuxpad /r/opengl-driver> ls /run/opengl-driver/lib/
d3d             libEGL_mesa.so.0      libGLX_mesa.so.0            libVkLayer_MESA_device_select.so  libvulkan_intel_hasvk.so  libvulkan_nouveau.so  libxatracker.so        vdpau
dri             libEGL_mesa.so.0.0.0  libGLX_mesa.so.0.0.0        libVkLayer_MESA_overlay.so        libvulkan_intel.so        libvulkan_radeon.so   libxatracker.so.2
libEGL_mesa.so  libGLX_mesa.so        libVkLayer_INTEL_nullhw.so  libvulkan_dzn.so                  libvulkan_lvp.so          libvulkan_virtio.so   libxatracker.so.2.5.0
```

```console
[rolf@tuxpad:~]$ cd $(dirname $(realpath $(which VirtualBox)))/../libexec/virtualbox

[rolf@tuxpad:/nix/store/jvk26bpi3gnak495gz30majywvirwcmw-virtualbox-7.1.6a/libexec/virtualbox]$ ldd *so | grep -B 3 libGL.so
VBoxSVGA3D.so:
        linux-vdso.so.1 (0x00007fffd531d000)
        VBoxRT.so => /nix/store/jvk26bpi3gnak495gz30majywvirwcmw-virtualbox-7.1.6a/libexec/virtualbox/./VBoxRT.so (0x00007fa6e4e00000)
        libGL.so.1 => /nix/store/kynlrr98p2c235b00c72b14apn7l8l4y-libglvnd-1.7.0/lib/libGL.so.1 (0x00007fa6e52d5000)
```

### System metadata

```
[I] rolf@tuxpad ~/s/opengl-virtualbox> nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.76, NixOS, 24.11 (Vicuna), 24.11.20250210.44534bc`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.12`
 - channels(root): `"nixos-24.05"`
 - channels(rolf): `""`
 - nixpkgs: `/nix/store/z88hhaq46sdqzkm0zas1sn284h7w87k9-source`

```

### Notify maintainers

@blitz You might be the right person to look into this.

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+bug%22+-label%3A%226.topic%3A+darwin%22+-label%3A%226.topic%3A+nixos%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/applications/virtualization/virtualbox/default.nix
+++ b/pkgs/applications/virtualization/virtualbox/default.nix
@@ -49,7 +49,6 @@
   open-watcom-bin,
   makeself,
   perl,
-  vulkan-loader,
   javaBindings ? true,
   jdk, # Almost doesn't affect closure size
   pythonBindings ? false,
@@ -163,11 +162,11 @@ stdenv.mkDerivation (finalAttrs: {
     libtpms
     python3
     xz
+    libGL
   ]
   ++ optional javaBindings jdk
   ++ optional pythonBindings python3 # Python is needed even when not building bindings
   ++ optional pulseSupport libpulseaudio
-  ++ optionals headless [ libGL ]
   ++ optionals (!headless) [
     qtbase
     qttools
@@ -212,6 +211,13 @@ stdenv.mkDerivation (finalAttrs: {
 
     grep 'libasound\.so\.2'     src include -rI --files-with-match | xargs sed -i -e '
       s@"libasound\.so\.2"@"${alsa-lib.out}/lib/libasound.so.2"@g'
+
+    substituteInPlace src/VBox/Devices/Graphics/DevVGA-SVGA3d-glLdr.cpp \
+      --replace-fail \"libGL.so.1\" \"${libGL.out}/lib/libGL.so.1\"
+
+    # this works in conjunction with fix-graphics-driver-loading.patch
+    substituteInPlace src/VBox/Devices/Graphics/DevVGA-SVGA3d-dx-dx11.cpp \
+      --replace-fail \"VBoxDxVk\" \"$out/libexec/virtualbox/VBoxDxVk.so\"
 
     export USER=nix
     set +x
@@ -260,6 +266,7 @@ stdenv.mkDerivation (finalAttrs: {
       ./qt-dependency-paths.patch
       # https://github.com/NixOS/nixpkgs/issues/123851
       ./fix-audio-driver-loading.patch
+      ./fix-graphics-driver-loading.patch
     ];
 
   postPatch = ''
@@ -394,8 +401,7 @@ stdenv.mkDerivation (finalAttrs: {
     # If hardening is disabled, wrap the VirtualBoxVM binary instead of patching
     # the source code (see postPatch).
     + optionalString (!headless && !enableHardening) ''
-      wrapQtApp $out/libexec/virtualbox/VirtualBoxVM \
-         --prefix LD_LIBRARY_PATH : "${lib.makeLibraryPath [ vulkan-loader ]}"
+      wrapQtApp $out/libexec/virtualbox/VirtualBoxVM
     '';
 
   passthru = {

--- a/pkgs/applications/virtualization/virtualbox/fix-graphics-driver-loading.patch
+++ b/pkgs/applications/virtualization/virtualbox/fix-graphics-driver-loading.patch
@@ -1,0 +1,13 @@
+diff --git a/src/VBox/Devices/Graphics/DevVGA-SVGA3d-dx-dx11.cpp b/src/VBox/Devices/Graphics/DevVGA-SVGA3d-dx-dx11.cpp
+index 1a43382..c376d6e 100644
+--- a/src/VBox/Devices/Graphics/DevVGA-SVGA3d-dx-dx11.cpp
++++ b/src/VBox/Devices/Graphics/DevVGA-SVGA3d-dx-dx11.cpp
+@@ -3376,7 +3376,7 @@ static DECLCALLBACK(int) vmsvga3dBackInit(PPDMDEVINS pDevIns, PVGASTATE pThis, P
+     AssertReturn(pBackend, VERR_NO_MEMORY);
+     pThisCC->svga.p3dState->pBackend = pBackend;
+ 
+-    rc = RTLdrLoadSystem(VBOX_D3D11_LIBRARY_NAME, /* fNoUnload = */ true, &pBackend->hD3D11);
++    rc = RTLdrLoad(VBOX_D3D11_LIBRARY_NAME, &pBackend->hD3D11);
+     AssertRC(rc);
+     if (RT_SUCCESS(rc))
+     {

